sourcemap/store.go

```diff
@@ -44,39 +41,103 @@ var (
 
 // Store holds information necessary to fetch a sourcemap, either from an Elasticsearch instance or an internal cache.
 type Store struct {
-	cache   *gocache.Cache
-	esStore *esStore
-	logger  *logp.Logger
+	cache       *gocache.Cache
+	backend     backend
+	logger      *logp.Logger
+	waiters     chan struct{}
+	waitTimeout time.Duration
+
+	sync.Mutex
+	inflight map[string]chan struct{}
+}
+
+type backend interface {
+	fetch(ctx context.Context, name, version, path string) (string, error)
 }
 
-// NewStore creates a new instance for fetching sourcemaps. The client and index parameters are needed to be able to
-// fetch sourcemaps from Elasticsearch. The expiration time is used for the internal cache.
-func NewStore(client elasticsearch.Client, index string, expiration time.Duration) (*Store, error) {
-	if expiration < 0 {
+func newStore(
+	b backend,
+	logger *logp.Logger,
+	cacheExpiration, waitTimeout time.Duration,
+	maxWaiters int,
+) (*Store, error) {
+	if cacheExpiration < 0 {
 		return nil, errInit
 	}
-	logger := logp.NewLogger(logs.Sourcemap)
+
 	return &Store{
-		cache:   gocache.New(expiration, cleanupInterval(expiration)),
-		esStore: &esStore{client: client, index: index, logger: logger},
+		cache:   gocache.New(cacheExpiration, cleanupInterval(cacheExpiration)),
+		backend: b,
 		logger:  logger,
+		// TODO: Number of requests waiting for a sourcemap to be
+		// fetched. Any suggestion as to where we should set this is
+		// appreciated.
+		waiters:     make(chan struct{}, maxWaiters),
```

Comment:
I think protecting against request storms with the inflight map is a good idea, but I'm not sure about this waiters limit. What is it protecting against?

Reply:
With the inflight map we'll only send a single request to the fleet server, but if the fleetserver is down or slow, then we'll start piling up requests to apm-server who are waiting for the result of fetch, which could potentially bring down the apm-server. Without having checked any numbers this could definitely fall into the realm of being overly-defensive, so no worries if you want me to drop it.

Reply:
I'd prefer to drop it, or at least remove it from this PR.

Source mapping is (at least for now) initiated by the "publisher" (apm-server/publish), which has a worker per GOMAXPROCS -- one per CPU. Thus there shouldn't ever be a large backlog, unless we were to shuffle source mapping to somewhere else in the process.

Aside from that we already have other safeguards in place: rate-limiting for RUM, and it is possible to set a maximum number of concurrent connections.